_posts/2021-03-14-Julia-1.6-what-has-changed-since-1.0.md

```diff
@@ -0,0 +1,737 @@
+For [full details see the original issue](https://github.com/JuliaLang/Pkg.jl/issues/1377) that introduced the Pkg protocol, it has a ton of info at the top explaining how it all works.
+There are [distributed PkgServers around the world](https://status.julialang.org/).
+
+Not only does this solve the latency problems of connecting to some GitHub server that is on the other side of the world, it also improves support for non-GitHub hosting.
+Its always been possible to host Julia packages on anything supporting git.
+There are [over a dozen packages in General Registry hosted on GitLab](https://github.com/JuliaRegistries/General/search?q=gitlab) for example.
+My employer, [Invenia](https://www.invenia.ca/), has all our internal packages hosted on GitLab, and registered on our internal registry.
+But in effect, prior to PkgServer, GitHub was 1st class in Pkg, due to its download as a tarball feature.
+It makes sense, GitLab and BitBucket don't even provide an API for downloading a repo via content (tree) hash (only commit hash), so it can't be implemented inside Pkg itself.
+But by going via PkgServer, everything can be consistently served.
+
+PkgServer has not been without struggles.
+Both social and technical.
+Early versions of the system stored some metrics, so that package authors could be informed of how many users their package had, what operating systems they used etc.
+After much debate, it was concluded that this might be too invasive, and it was removed before the Julia 1.5 release.
+On the technical side, it initially had some struggles with keeping up with the load, with some servers becoming overloaded.
+More recently its had some issues (hopefully now fixed) where it has not been able to see the latest releases of packages because it is still seeing an out-of-date copy of the registry.
+
+As [Jeff Atwood said](https://twitter.com/codinghorror/status/506010907021828096):
+
+> There are two hard things in computer science: cache invalidation, naming things, and off-by-one errors.
+
+[![XKCD 1854](https://imgs.xkcd.com/comics/refresh_types.png)
+(XKCD 1845](https://xkcd.com/1854/))
+
+## Time To First Plot
+People often complain about the "Time To First Plot" (TTFP) in Julia.
+I personally have never minded it -- by the time I am plotting something, I have done minutes of thinking so 20 seconds of compilation is nothing.
+Plotting, it turns out, is basically a really hard thing for a compiler.
+It is many, many, small methods, most of which are only called once.
+And unlike most Julia code, it doesn't actually benefit all that much from Julia's JIT.
+Julia's JIT is normally specializing code, and running a ton of optimizations.
+But plotting itself isn't in the hot-loop -- optimizing the code takes longer than running it the few dozen times it might be used unoptimized.
+To make a long-story short, plotting is the postcard example for Julia needing to compile things before it can run them.
```

Comment:
postcard example -> prime example

Reply:
(Unless I'm wholly unfamiliar with this idiom, which ESL, but I don't think it exists.)

Reply:
I've heard "poster child", but not "postcard"

Reply:
```suggestion
To make a long-story short, plotting is the poster child example for Julia needing to compile things before it can run them.
```